src/naive/datetime/mod.rs

```diff
@@ -131,6 +131,101 @@ impl NaiveDateTime {
         datetime.expect("invalid or out-of-range datetime")
     }
 
+    /// Makes a new `NaiveDateTime` corresponding to a UTC date and time,
+    /// from the number of non-leap milliseconds
+    /// since the midnight UTC on January 1, 1970 (aka "UNIX timestamp")
+    ///
+    /// Returns `None` on the out-of-range number of milliseconds.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use chrono::{NaiveDateTime, NaiveDate};
+    ///
+    ///
+    /// let dt = NaiveDate::from_ymd(1970, 1, 1).and_hms_milli(0, 0, 0, 42);
+    ///
+    /// let got = NaiveDateTime::from_timestamp_millis(dt.timestamp_millis()).unwrap();
+    /// assert_eq!(dt, got);
+    /// let dt = NaiveDate::from_ymd(3015, 9, 5).and_hms(23, 56, 4);
+    /// let ms = dt.timestamp_millis();
+    /// assert_eq!(NaiveDateTime::from_timestamp_millis(ms), Some(dt));
+    /// ```
+    #[inline]
+    pub fn from_timestamp_millis<T: Into<i64>>(milliseconds: T) -> Option<NaiveDateTime> {
+        let milliseconds = milliseconds.into();
+        let timestamp_seconds = milliseconds / 1000;
+        NaiveDateTime::from_timestamp_opt(
+            timestamp_seconds,
+            ((milliseconds.abs() % 1000) * 1_000_000) as u32,
```

Comment:
I don't think this `abs()` call is correct. If `milliseconds` is -400, we end up with a `NaiveDateTime` that is at 400_000_000ns *after* the epoch, which is clearly wrong. Should probably add a test (a normal test please, not a doc test) for this.

Same for the other new methods.